apiserver/facades/client/application/application.go

```diff
@@ -2199,3 +2288,81 @@ func (api *APIBase) ApplicationsInfo(in params.Entities) (params.ApplicationInfo
 	}
 	return params.ApplicationInfoResults{out}, nil
 }
+
+// lxdCharmProfiler massages a *state.Charm into a LXDProfiler
+// inside of the core package.
+type lxdCharmProfiler struct {
+	Charm Charm
+}
+
+// LXDProfile implements core.lxdprofile.LXDProfiler
+func (p lxdCharmProfiler) LXDProfile() lxdprofile.LXDProfile {
+	if p.Charm == nil {
+		return nil
+	}
+	if profiler, ok := p.Charm.(charm.LXDProfiler); ok {
+		return profiler.LXDProfile()
+	}
+	return nil
+}
+
+// ModelConfig is a point of use model configuration.
+type ModelConfig interface {
+	AgentVersion() (version.Number, bool)
+}
+
+var (
+	// ErrInvalidAgentVersions is a sentinal error for when we can no longer
+	// upgrade juju using 2.5.x agents with 2.6 or greater controllers.
+	ErrInvalidAgentVersions = errors.Errorf(
+		"Unable to upgrade LXDProfile charms with the current model version. " +
+			"Please run juju upgrade-juju to upgrade the current model to match your controller.")
+)
+
+func validateAgentVersions(application Application, cfgAgentVersion ModelConfig) error {
+	var agentVer version.Number
+	var noAgentVersion bool
+
+	// The epoch is set like this, because beta tags are less than release tags.
+	// So 2.6-beta1.1 < 2.6.0, even though the patch is greater than 0. To
+	// prevent the miss-match, we add the upper epoch limit.
+	epoch := version.Number{Major: 2, Minor: 5, Patch: math.MaxInt32}
```

Comment:
Does Version.Number.Compare(other.Number) not work?

If it doesn't work, this code should be factored out into something that just compares versions, rather than co-mingling the version comparison with loading what the versions/tools/agents are.